### ISSUE TITLE
why char[] in reloption struct?

### ISSUE THREAD

**Issue by kainwen** [greenplum-db/gpdb#5139]
In the code here https://github.com/greenplum-db/gpdb/blob/master/src/include/utils/rel.h#L267

```c
typedef struct StdRdOptions
{
	int32		vl_len_;		/* varlena header (do not touch directly!) */
	int			fillfactor;		/* page fill factor in percent (0..100) */
	AutoVacOpts autovacuum;		/* autovacuum-related options */

	bool		appendonly;		/* is this an appendonly relation? */
	int			blocksize;		/* max varblock size (AO rels only) */
	int			compresslevel;  /* compression level (AO rels only) */
	char		compresstype[NAMEDATALEN]; /* compression type (AO rels only) */
	bool		checksum;		/* checksum (AO rels only) */
	bool 		columnstore;	/* columnstore (AO only) */
	char		orientation[NAMEDATALEN]; /* orientation (AO only) */
} StdRdOptions;

```

Add two char [] field in the struct `StdRdOptions`. It is introduced in the commit:  https://github.com/greenplum-db/gpdb/commit/25a90396cd1c52d252a37986e12b63e8e037aa83

The `compresstype` field  and `orientation` field are set in by the code in the function `fillRelOptions `, the code is here:  https://github.com/greenplum-db/gpdb/blob/master/src/backend/access/common/reloptions.c#L1098

```c
					case RELOPT_TYPE_STRING:
						optstring = (relopt_string *) options[i].gen;
						if (options[i].isset)
							string_val = options[i].values.string_val;
						else if (!optstring->default_isnull)
							string_val = optstring->default_val;
						else
							string_val = NULL;

						if (string_val == NULL)
							*(int *) itempos = 0;
						else
						{
							strcpy((char *) rdopts + offset, string_val);
							*(int *) itempos = offset; // <----------Look at the line
							offset += strlen(string_val) + 1;
						}
						break;

``` 

The code is from upstream(pg). Look at this line of code `*(int *) itempos = offset;`

`itempos` is pointed to `StdRdOptions->compresstype`. Why we convert a `char (*)[]` to `int *`, and store an `int` into the field?  **It is much too confusing**.

In fact, postgres do not have string_type reloption. Greenplum has string_type reloptions only for AO table. And Greenplum rewrites it in `reloption_gp.c`.

The `char compresstype[NAMEDATALEN]` and https://github.com/greenplum-db/gpdb/blob/master/src/backend/access/common/reloptions.c#L1098 are dead code.

I've no good idea on how to fix(modify) it. But define these two fields as `char[]` and store `int` in them, is not correct I believe.

**Comment by kainwen** [comment]
@asimrp

**Comment by d** [comment]
Zhenghua you are right!

**Comment by asimrp** [comment]
Good catch, @kainwen The right way to fix, according to @dgkimura and I, is to convert `compresstype` and `orientation` to `enum` rather than strings.  Note that although what we have is  confusing to read, it is not a bug.  Nothing is breaking because all the AO specific options are overwritten right after `fillRelOptions()` call in `validate_and_refill_options()`.

**Comment by kainwen** [comment]
@asimrp 
Yes. It is not a bug because we never use that field.

Recently I spike to add some reloptions and this confuses me. So I post it here.

**Comment by yydzero** [comment]
@asimrp shall we fix this issue or won't fix it?

**Comment by asimrp** [comment]
The code is fragile and waiting for bugs to be introduced.  It needs to be fixed.

**Comment by ashwinstar** [comment]
@huansong or @bmdoil given you have looked into this area for other things, wish to look into this and fix it out if still the case.

**Comment by huansong** [comment]
This is still the case for both 6X and 7X. But 7X doesn't have `orientation` anymore (it's not a reloption now but recognized as part of the table AM). So it is just `compresstype`.

The weird way of copying an integer into `char*` field is meant to (I think) copy varied lengths of strings. The string will be copied over at the end of `StdRdOptions`, with a position `offset` stored in the `char*` field. And the user of `StdRdOptions` will know where to get the string. But this design is currently not used by GPDB because we store a full-length `compresstype` in `StdRdOptions`:
```
char		compresstype[NAMEDATALEN];
```
and we copy to and from this `compresstype` directly. 
If we want to utilize this design, we'll need to at least do these things:
* change the `compresstype` field in `StdRdOptions` to an `int` because as mentioned by @kainwen  using `char*` to store offset is super confusing;
* change how `compresstype` is being used currently, [for example](https://github.com/greenplum-db/gpdb/blob/main/src/backend/catalog/pg_appendonly.c#L126), to use the data portion indicated by the offset.
* do not refill or adjust `compresstype` in `validate_and_adjust_options`. Move that logic prior to allocating `StdRdOptions` (`allocateReloptStruct`), because the string length could change after the refill/adjust, and we need the "final" string length to be able to allocate and copy properly.

These are not hard to do but the question is what's the benefit. As mentioned this is not a bug, we just have special handling of AO reloptions as a whole which is not limited to a string option. So doing the above alone probably would make the code even more confusing. I think we should make this as part of complete refactoring of the special handling of AO reloptions (we did some in #14850 but not enough). We have a backlog story for that.
